benchmarks/instruction_counts/execution/work.py

```diff
@@ -0,0 +1,204 @@
+"""Handle the details of subprocess calls and retries for a given benchmark run."""
+import dataclasses
+import os
+import pickle
+import signal
+import subprocess
+import time
+from typing import List, Optional, Union, TYPE_CHECKING
+import uuid
+
+from core.api import AutoLabels
+from core.types import Label
+from core.utils import get_temp_dir
+from worker.main import WORKER_PATH, WorkerFailure, WorkerOutput, WorkerTimerArgs, WorkerUnpickler
+
+if TYPE_CHECKING:
+    PopenType = subprocess.Popen[bytes]
+else:
+    PopenType = subprocess.Popen
+
+
+# Mitigate https://github.com/pytorch/pytorch/issues/37377
+_ENV = "MKL_THREADING_LAYER=GNU"
+_PYTHON = "python"
+PYTHON_CMD = f"{_ENV} {_PYTHON}"
+
+# We must specify `bash` so that `source activate ...` always works
+SHELL = "/bin/bash"
+
+
+@dataclasses.dataclass(frozen=True)
+class WorkOrder:
+    """Spec to schedule work with the benchmark runner."""
+    label: Label
+    autolabels: AutoLabels
+    timer_args: WorkerTimerArgs
+    source_cmd: Optional[str] = None
+    timeout: Optional[float] = None
+    retries: int = 0
+
+    def __hash__(self) -> int:
+        return id(self)
+
+
+class _BenchmarkProcess:
+    """Wraps subprocess.Popen for a given WorkOrder."""
+    _work_order: WorkOrder
+    _cpu_list: Optional[str]
+    _proc: PopenType
+
+    # Internal bookkeeping
+    _communication_file: str
+    _start_time: float
+    _end_time: Optional[float] = None
+    _retcode: Optional[int]
+    _result: Optional[Union[WorkerOutput, WorkerFailure]] = None
+
+    def __init__(self, work_order: WorkOrder, cpu_list: Optional[str]) -> None:
+        self._work_order = work_order
+        self._cpu_list = cpu_list
+        self._start_time = time.time()
+        self._communication_file = os.path.join(get_temp_dir(), f"{uuid.uuid4()}.pkl")
+        with open(self._communication_file, "wb") as f:
```

Comment:
The communication file thing makes me a little queazy. A few thoughts here:
Can you used NamedTemporaryFile? Are we sure we don't need flock (I suppose the life cycle of the subprocess keeps it safe)? Is standard input & standard output not appropriate for passing this data?

Reply:
Yeah, in general I'm relying on lifetimes for safety. The process doesn't get launched until the input is written, and the output isn't read until the child has terminated. So I don't think I need flock, though if I'm wrong by all means let me know. 

The reason I don't have to use `NamedTemporaryFile` is that `_communication_file` lives in `get_temp_dir()`, and that dir is guarded by an `atexit.register(shutil.rmtree, ...)`, so rather than worrying after individual files I just guaranteed that hell or high water everything gets cleaned up. (Except something like SIGKILL, of course. But I can't really help that...) I don't like using stdin/stdout because the Python PIPE implementation has as fixed 16k size and is prone to hang if that is exceeded. (Plus I don't want to pull pickle dumps out of a stream...)

That said, this is just my thought process. It is entirely possible that I'm violating some best practices. I'm going to leave this unresolved until you're satisfied.